### ISSUE TITLE
Unsupported -arch flag on Darwin (migrated from Trac #1087)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2640]
Original ticket http://projects.scipy.org/numpy/ticket/1087
Reported 2009-04-16 by trac user smaret, assigned to atmention:dmcooke.

Numpy 1.3.0 uses a "-arch" flag for the gfortran compiler on Darwin. This flag is only supported by universal-capable compilers, which is not the case for the standard gfortran from gcc 4.x. This breaks the compilation of modules that uses a fortran compiler, such as scipy.

There is a function (_can_target) in numpy/distutils/fcompiler/gnu.py that supposedly tests if the compiler support the -arch flag, but apparently its is not working:

```
% python -c 'from numpy.f2py.diagnose import run; run()' | grep -A 10 "Gnu95FCompiler"
Gnu95FCompiler instance properties:
  archiver        = ['/sw/bin/gfortran', '-cr']
  compile_switch  = '-c'
  compiler_f77    = ['/sw/bin/gfortran', '-Wall', '-ffixed-form', '-fno-
                    second-underscore', '-arch', 'i686', '-fPIC', '-O3', '-
                    funroll-loops']
  compiler_f90    = ['/sw/bin/gfortran', '-Wall', '-fno-second-underscore',
                    '-arch', 'i686', '-fPIC', '-O3', '-funroll-loops']
  compiler_fix    = ['/sw/bin/gfortran', '-Wall', '-ffixed-form', '-fno-
                    second-underscore', '-Wall', '-fno-second-underscore', '-
                    arch', 'i686', '-fPIC', '-O3', '-funroll-loops']
```

I am using gfortran 4.3.3 and Python 2.5.4 installed with Fink on MacOS 10.5.6.

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-04-20

What does gfortran -v and gfortran -arch ppc -v do with your gfortran version ? I don't have fink installed and would prefer not installing it, but I build my own non universal gcc from pristine gnu sources, and the _can_target function works as expected. IOW, with a non universal gfortran, the -arch foo flags are not added.

**Comment by thouis** [comment]
Comment in Trac by trac user smaret, 2009-04-20

Here what I obtain:

```
% gfortran -v
Using built-in specs.
Target: i686-apple-darwin9
Configured with: ../gcc-4.3.3/configure --prefix=/sw --prefix=/sw/lib/gcc4.3 --mandir=/sw/share/man --infodir=/sw/share/info --enable-languages=c,c++,fortran,objc,java --with-arch=nocona --with-tune=generic --build=i686-apple-darwin9 --with-gmp=/sw --with-libiconv-prefix=/sw --with-system-zlib --x-includes=/usr/X11R6/include --x-libraries=/usr/X11R6/lib --disable-libjava-multilib
Thread model: posix
gcc version 4.3.3 (GCC) 
% gfortran -arch pcc -v
Using built-in specs.
Target: i686-apple-darwin9
Configured with: ../gcc-4.3.3/configure --prefix=/sw --prefix=/sw/lib/gcc4.3 --mandir=/sw/share/man --infodir=/sw/share/info --enable-languages=c,c++,fortran,objc,java --with-arch=nocona --with-tune=generic --build=i686-apple-darwin9 --with-gmp=/sw --with-libiconv-prefix=/sw --with-system-zlib --x-includes=/usr/X11R6/include --x-libraries=/usr/X11R6/lib --disable-libjava-multilib
Thread model: posix
gcc version 4.3.3 (GCC)
```

When I try to compile a simple program with the -arch option, I get the following error:

```
% cat test.f

[gag8175:~] smaret% cat test.f90
program test

  write (*,*) "hello !"

end
% gfortran -arch pcc test.f90
f951: error: unrecognized command line option "-arch"
```

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-04-20

Ok, could you update to revision r6877, and try the following code ? You should have gfortran in your path, and the code should be executed from the numpy tree

```
#!python
import __builtin__
__builtin__.__NUMPY_SETUP__ = True

from numpy.distutils.fcompiler.gnu import _can_target
for arch in ['ppc', 'ppc64', 'i686', 'x86_64']:
    print "====== Can handle arch %s ? ========= " % _can_target(['gfortran'], arch)
```

If your gfortran is not universal, you should get False for every arch. If not, there is a bug in _can_target. If you get False for every arch, but you still get the -arch i686 added, then the flags come from something else.

**Comment by thouis** [comment]
Comment in Trac by trac user leek, 2009-04-23

Hi all,
Coincidentally, I am having the same problem for the same reason (fink installation of non-universal gfortran. I tried:

gfortran -arch i686 -v

which completes successfully.

However:

gfortran -arch i686 test.f
gives the error:
f951: error: unrecognized command line option "-arch"

I'd say that the bug isn't in Numpy, but I suppose the solution is to specify a Fortran file when executing the command.

Hope this helps. I may try to patch seeing as I can recreate the problem.

--Lee

**Comment by thouis** [comment]
Attachment in Trac by trac user leek, 2009-04-23: [gnu.py.patch](http://projects.scipy.org/numpy/attachment/ticket/1087/gnu.py.patch)

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-06-01

I think r6877 fixed it.

**Comment by thouis** [comment]
Comment in Trac by trac user leek, 2009-06-01

Replying to [comment:5 cdavid]:

> I think r6877 fixed it.
> It does look like r6877 addresses the problem. However, the code added by r6877 does not detect "-arch" support; apparently the compiler accepts the "-arch" switch when combined with the "-v" switch, but fails only when the compiler is fed a real file. My patch supplies the compiler with a real file which triggers the compiler failure and allows the code to detect lack of -arch support. Perhaps this is a bug in the compiler, but my patch does provide a work-around.

The problem occurred on my machine with a post-r6877 gnu.py. I would like you to include the patch. I realize that recreating the bug requires quite a lot of effort on your part. Perhaps if the patch otherwise does no harm, that is enough to argue for its inclusion?

I don't have authority to reopen this ticket but I believe it should be reopened.

Thanks,
--Lee

**Comment by thouis** [comment]
Comment in Trac by trac user fitz, 2009-07-13

I was bitten by this bug as well, in r7132.  I applied the patch supplied by leek and found that it was a successful fix.  I believe this patch (or something like it) should be included.

**Comment by thouis** [comment]
Comment in Trac by trac user fitz, 2009-10-28

This bug is still present in r7596.  This ticket should be reopened until something like Lee's patch (supplying a test file to test -arch support) is applied.

**Comment by thouis** [comment]
Comment in Trac by trac user grobian, 2009-11-20

Can I ask what was the rationale behind adding -arch support?  I cooked up a patch as well, removing the entire -arch addition, because IMO it simply doesn't make much sense to me.

Apple never released a fortran compiler, so how can universal support ever be expected for it to exist (and work correctly)?.  This is my gfortran compiler:

% gfortran --version
GNU Fortran (GCC) 4.2.1 (Apple Inc. build 5574)
Copyright (C) 2007 Free Software Foundation, Inc.

And indeed as mentioned before, it happily accepts any -arch argument with -v, but in real action it always complains.

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-11-25

The rationale is that the recommended gfortran _does_ support the -arch flag:

http://r.research.att.com/tools/

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-12-10

Fixed in r7990 for the 1.4.x branch, and r7991 in the trunk.

**Comment by thouis** [comment]
Comment in Trac by trac user nightwolf67, 2010-11-23

[http://www.x4labs.com/penis-pumps.php penis pump] [http://www.sinrex.com/male-enhancement-cream.php male enhancement cream]
